Node JSON support

sandstrom opened the pull request:
In chef nodes can be passed individual JSON snippets, called *node JSON*. Node JSON is used to customize certain things per node.

Some examples are host names, unique node ids, unique tokens for nodes to 'phone home' and report that they are setup successfully (for load-balancers registration, etc), adding per-node unique software license keys, data to seed the random generator and much more.

This commit adds node-JSON support to stemcell.

`jq` is like `sed` for JSON, is written in c and has no runtime dependencies. Packages exist in the official repositories for Debian, Ubuntu, Fedora, openSUSE, Arch and Solaris (https://stedolan.github.io/jq/). The footprint is small, it clocks in at ~50 kb.

@darnaut Hi Davi, I'm eager to hear your thoughts on this!

Closes #93

sandstrom commented:
friendly ping! 😄 @darnaut @ljharb @noggi

igor47 commented on lib/stemcell/templates/bootstrap.sh.erb:
```diff
@@ -48,6 +49,7 @@ domain_name='<%= opts['instance_domain_name'] %>'
 chef_version='<%= opts['chef_version'] %>'
 username='<%= opts.fetch('user', ENV['USER']) %>'
 encrypted_data_bag_secret_path='<%= opts['chef_data_bag_secret_path'].to_s %>'
+node_json='<%= opts['node_json'] %>'
```
the quoting would break if the node json contains `'` characters. this requires a more serious escaping strategy -- a HEREDOC might to the trick.

sandstrom replied:
Good point, I've added a commit that solves this.

igor47 commented on lib/stemcell/launcher.rb:
```diff
@@ -181,6 +182,14 @@ def launch(opts={})
         end
       end
 
+      if opts['node_json'].is_a?(Hash)
```
do you expect this to be a hash, or a string?

sandstrom replied:
Both, but I agree that it wasn't very clear. 

The old code would (1) turn a hash into JSON string, (2) allow a string (any string) and (3) if no string was set in step (1) or passed in, it would default to an empty JSON object.

I've now updated it to this, which is more robust:

```ruby
      if opts['node_json'].is_a?(Hash)
        opts['node_json'] = JSON.generate(opts['node_json'])
      elsif opts['node_json'].is_a?(String)
        unless valid_json?(opts['node_json'])
          raise '`node_json` must be a hash or a valid json string'
        end
      else # fallback
        opts['node_json'] = '{}'
      end
```

igor47 commented on lib/stemcell/templates/bootstrap.sh.erb:
```diff
@@ -226,7 +232,9 @@ GIT_SSH=${git_wrapper} git fetch
 
-json="{\"role\": \"\${role}\", \"env\": \"\${env}\", \"branch\": \"\${branch}\"}"
+base_json="{\"role\": \"\${role}\", \"env\": \"\${env}\", \"branch\": \"\${branch}\"}"
+json=\`echo \$node_json \$base_json | jq -s '.[0] + .[1]'\`
```
there's no reason to involve an external binary; you can just construct the variable you want in ruby and pass it, fully constructed, into the template.

sandstrom replied:
Alright, I've added a commit which moves this over to 'ruby-land'.

sandstrom commented:
@igor47 First of all, thanks for reviewing! 😄 Much appreciated!

I've addressed your points in new commits. Let me know if you have more comments or suggestions.

Regarding the use-case, I'd say there are a few:

- License keys
- Custom hostnames
- Post setup phone home (will usually require a node-unique identifier)
- Other custom identifiers, for example in a Consul cluster or for databases in a replication ring/cluster.

sandstrom commented:
@igor47 I've run across an issue related to node-data, but for environment.

On `master` this tool set the environment via a `env` key in the custom-json passed to the node.
Code here: https://github.com/airbnb/stemcell/blob/master/lib/stemcell/templates/bootstrap.sh.erb#L229

Is this an old way of setting the env in chef, or something custom? Afaik it should be set via the `-E` flag passed to `chef-solo` ([docs here](https://docs.chef.io/chef_solo.html)).

Also, the runlist/role/env dance, is that something custom?

```
runlist="role[\${role}]"
if [ -f "roles/\${env}.rb" ]
then
  runlist+=",role[\${env}]"
fi
```

sandstrom commented:
@igor47 friendly ping 😄

sandstrom commented:
Hey @igor47, do you have a few minutes to look at this?
